File: apps/ai-service/app/routers/skills.py
"""Skill taxonomy seams: POST /skills/canonicalize, POST /skills/retag-plan."""

from __future__ import annotations

import asyncio

from fastapi import APIRouter

from ..ai import cost_tracker
from ..ai.canonicalize import canonicalize_skill
from ..ai.embeddings import EMBEDDING_TASK_TYPE
from ..ai.langfuse_tracing import get_tracer
from ..ai.model_config import rate_inr_per_1k
from ..ai.retag import plan_retag
from ..ai.skill_store import get_skill_store
from ..config import get_settings
from ..contracts import (
    RetagPlanInput,
    RetagPlanOutput,
    SkillCanonicalization,
    SkillCanonicalizationInput,
)
from ._shared import logger

api_router = APIRouter()


@api_router.post("/skills/canonicalize", response_model=SkillCanonicalization)
async def skills_canonicalize(body: SkillCanonicalizationInput) -> SkillCanonicalization:
    """ADR-0030 / TAX-6: the JOB side canonicalizes through the SAME pipeline as the
    worker side — one shared id space. The NestJS api calls this at job-posting
    create/update for each posting skill phrase; `canonicalize_skill` runs
    pseudonymize -> embed (SG-2/SG-4) -> scope-bounded nearest-alias (seam A store) ->
    floor gate. SG-3 holds: the id can only come from the closed skill_alias set.

    SCOPE (Phase 1.5): the body carries EXACTLY ONE of `domain_id` (legacy slug) or
    `job_domain_id` (canonical `jd_*`, resolved through `job_domain_skill` so a skill
    with a NULL legacy domain is reachable at all). Neither/both is a 422 from the
    contract — an unscoped search is refused, never widened.

    Honors SKILL_CANONICALIZE_ENABLED: flag off -> UNRESOLVED (inert — rollback for the
    job side is the same single flag as the worker side). ``async def`` + ``to_thread``:
    the store + a real embed are SYNC httpx calls and must not block the event loop
    (same posture as the #222 fix).

    TD68 (TD27 SpendLedger wiring, REAL embed path only): the single-phrase projected
    cost is reserved BEFORE the embed and reconciled after. A ledger block returns
    UNRESOLVED — canonicalization NEVER blocks the caller (the same posture as every
    other failure). Mock embed path (``real_call_enabled_for`` false): zero ledger
    traffic. Never logs the phrase.

    TRACED as one task with the embed nested inside it. Both non-embedding outcomes —
    the flag being off and a ledger block — are traced too, and deliberately: all three
    return the SAME ``unresolved`` body, so without the trace saying which one happened
    a disabled flag is indistinguishable from a phrase nothing matched. The phrase
    itself is never traced (the ``embed-text`` child records the MASKED text, which is
    what the provider actually saw)."""
    settings = get_settings()
    tracer = get_tracer()
    with tracer.task(
        task_type="skill_canonicalization",
        # BOTH SCOPE KEYS ARE TRACED, and both are safe: a legacy slug and a `jd_*` id are
        # closed-set identifiers, never worker text. Tracing only the one that was set would
        # make the two Phase 1.5 paths indistinguishable in Langfuse — the exact question the
        # cutover has to be able to answer ("did this go through job_domain_skill or the
        # legacy column?"). The phrase itself is still never traced.
        input={
            "domain_id": body.domain_id,
            "job_domain_id": body.job_domain_id,
            "lang": body.lang,
        },
        real_call=settings.real_call_enabled_for(EMBEDDING_TASK_TYPE),
    ) as task:
        if not settings.skill_canonicalize_enabled:
            task.update(level="WARNING", status_message="skill_canonicalize_disabled")
            return SkillCanonicalization(status="unresolved")

        embed_is_mock = not settings.real_call_enabled_for(EMBEDDING_TASK_TYPE)
        reserved_inr = 0.0
        if not embed_is_mock:
            in_rate, _out = rate_inr_per_1k(settings.embedding_model)
            reserved_inr = (cost_tracker.estimate_tokens(body.phrase) / 1000.0) * in_rate
            reason = await cost_tracker.get_ledger().would_exceed_spend(reserved_inr, settings)
            if reason is not None:
                # Counts/reason only — never the phrase.
                logger.warning(
                    "skills canonicalize blocked by spend ledger",
                    extra={"extra": {"reason": reason, "projected_inr": round(reserved_inr, 6)}},
                )
                task.update(level="WARNING", status_message=reason)
                return SkillCanonicalization(status="unresolved")

        actual_inr = 0.0
        try:
            out = await asyncio.to_thread(
                canonicalize_skill,
                body.phrase,
                body.domain_id,
                get_skill_store(settings),
                settings,
                lang=body.lang,
                # Phase 1.5: EXACTLY ONE of these is set (the contract 422s anything else),
                # so the legacy call is unchanged — this is None on every existing caller.
                job_domain_id=body.job_domain_id,
            )
            # Leave the full reservation recorded as the actual spend (conservative:
            # one real embed ran; a pseudonymize-blocked phrase slightly over-records).
            actual_inr = reserved_inr
            # SG-3 holds on the wire, so this is safe to record verbatim: `skill_id` can
            # only be an id from the closed skill_alias set, never model-invented text.
            task.update(output=out.model_dump())
            return out
        finally:
            if not embed_is_mock:
                # On a raise, actual_inr stayed 0.0 => full refund (no reservation leak).
                await cost_tracker.get_ledger().record_spend(reserved_inr, actual_inr)


@api_router.post("/skills/retag-plan", response_model=RetagPlanOutput)
def skills_retag_plan(body: RetagPlanInput) -> RetagPlanOutput:
    """ADR-0030 / TAX-9: compute the OFFLINE re-tag plan for deprecated skill ids.
    PURE COMPUTE — no LLM, no DB, no PII (row_refs are opaque uuids; ids are closed-set).
    The db-side runner (packages/db/src/retag-skills.ts, owner connection) supplies the
    ``skill.replaced_by`` crosswalk + the affected rows and APPLIES the plan only under
    ``--apply`` after a human reads the dry-run report. Chains resolve to the terminal
    id; cycles are dropped fail-safe (SG-5: ids immutable, plan never invents one).
    Same internal-only exposure posture as every ai-service route (TD67)."""
    out = plan_retag(body)
    logger.info(
        "skills retag plan",
        extra={
            "extra": {
                "crosswalk": len(body.crosswalk),
                "rows_in": out.rows_in,
                "rows_changed": out.rows_changed,
                "dropped_cyclic": len(out.dropped),
            }
        },
    )
    return out

File: packages/ai-contracts/src/skills.ts
import { z } from "zod";

import { AICallMetadataSchema } from "./common";

// Skill taxonomy seams (ADR-0030): phrase canonicalization (TAX-4), the alias
// embedding batch (TAX-3), growth-loop clustering (TAX-7) and the offline
// skill re-tag plan (TAX-9).

// ---------------------------------------------------------------------------
// Skill canonicalization (ADR-0030 / TAX-4) — mirrors contracts.py
// ---------------------------------------------------------------------------
/**
 * TWO DOMAIN ID SPACES, EXACTLY ONE PER REQUEST (Phase 1.5 canonicalizer cutover).
 *
 * `domain_id` is the LEGACY 11-slug skill domain ("cnc-machining") that
 * `skill_alias.domain_id` denormalizes. Migration 0076 made that column NULLABLE and
 * demoted it to compatibility metadata, so a canonical skill minted by the taxonomy
 * bootstrap carries no slug at all and is INVISIBLE to a `WHERE domain_id = $1` scan.
 * `job_domain_id` is the canonical `jd_*` id, and it resolves candidates through
 * `job_domain_skill` instead — which is the whole point of the cutover.
 *
 * EXACTLY ONE, ENFORCED — neither is 400, both is 400. This is the rule that matters
 * most in the file: making both optional without the refine would let a caller send
 * neither, and "no domain" must never be allowed to degrade into "search the entire
 * skill vocabulary". A cook would get offered lathe skills. Reject instead.
 *
 * A `.refine` (i.e. a ZodEffects, not a ZodObject) is safe here because nothing calls
 * `.shape`, `.extend`, `.partial` or `.pick` on this schema — verified at cutover; if
 * that ever changes, move the rule to `.superRefine` on the object and re-check.
 */
export const SkillCanonicalizationInputSchema = z
  .object({
    phrase: z.string(),
    /** LEGACY 11-slug skill domain. Optional as of Phase 1.5; still fully supported. */
    domain_id: z.string().optional(),
    /** CANONICAL `jd_*` job domain — candidates come from `job_domain_skill`. */
    job_domain_id: z.string().optional(),
    lang: z.string().default("en"),
  })
  .refine((v) => (v.domain_id === undefined) !== (v.job_domain_id === undefined), {
    message:
      "exactly one of domain_id (legacy skill-domain slug) or job_domain_id (canonical jd_*) is required",
  });
export type SkillCanonicalizationInput = z.infer<typeof SkillCanonicalizationInputSchema>;

// Result: an ASSIGNED skill_id (top match >= floor) or UNRESOLVED. No PII. SG-3 /
// LLM-never-invents: skill_id is null unless the vector layer assigned it.
export const SkillCanonicalizationSchema = z.object({
  status: z.enum(["matched", "unresolved"]),
  skill_id: z.string().nullable().default(null),
  score: z.number().nullable().default(null),
  // #745 — the embed is a PAID provider call and this contract had no way to carry its
  // cost, which is the same root cause #738 fixed for STT: no field ⇒ no emitter could
  // exist ⇒ `WHERE task_type = 'skill_embedding'` returned empty and read as "no spend".
  // Additive + defaulted, so an ai-service that predates the field still parses.
  // null on every path that reached no provider (flag off, ledger block, pseudonymize
  // block); `AiCostRecorder.record` no-ops on null so the caller never branches.
  ai_metadata: AICallMetadataSchema.nullable().default(null),
});
export type SkillCanonicalization = z.infer<typeof SkillCanonicalizationSchema>;

// ---------------------------------------------------------------------------
// Skill-alias embedding batch (ADR-0030 / TAX-3 fork-B runner seam) — mirrors
// contracts.py. The db-side runner (packages/db embed-skill-aliases.ts) POSTs
// alias-text batches to the ai-service /embeddings/skill-alias endpoint.
// ---------------------------------------------------------------------------
export const SkillAliasEmbedItemSchema = z.object({
  alias_id: z.string(),
  text: z.string(),
});
export type SkillAliasEmbedItem = z.infer<typeof SkillAliasEmbedItemSchema>;

export const SkillAliasEmbedInputSchema = z.object({
  items: z.array(SkillAliasEmbedItemSchema).max(200), // request cap == Pydantic max_length
});
export type SkillAliasEmbedInput = z.infer<typeof SkillAliasEmbedInputSchema>;

// vector null ⇔ blocked (pseudonymize fail-closed) — the runner leaves that row NULL.
export const SkillAliasEmbedResultSchema = z.object({
  alias_id: z.string(),
  vector: z.array(z.number()).nullable().default(null),
  blocked: z.boolean().default(false),
});
export type SkillAliasEmbedResult = z.infer<typeof SkillAliasEmbedResultSchema>;

// `results` may be SHORTER than `items`: budget-stopped or provider-errored items are
// OMITTED (rows stay NULL; a later run resumes). Already-paid embeds are always returned.
export const SkillAliasEmbedOutputSchema = z.object({
  results: z.array(SkillAliasEmbedResultSchema),
  is_mock: z.boolean().default(true),
  model: z.string(),
  // Per-request INR ceiling fired on the REAL path (TD64 interim guard).
  budget_stopped: z.boolean().default(false),
  // Per-item real-provider failures skipped (batch continued).
  errors: z.number().int().nonnegative().default(0),
  estimated_cost_inr: z.number().nonnegative().default(0),
});
export type SkillAliasEmbedOutput = z.infer<typeof SkillAliasEmbedOutputSchema>;

// ---------------------------------------------------------------------------
// Growth-loop clustering (ADR-0030 / TAX-7 — pure compute, human-gated) — mirrors
// contracts.py. The db-side runner (packages/db growth-cluster.ts) POSTs per-domain
// batches of OPEN unresolved_phrase rows (SG-1 pseudonymized text + vectors) and the
// embedded skill_alias anchors to /growth/cluster; the output is REPORT-ONLY — the
// human ratification flow is the only activation path.
// ---------------------------------------------------------------------------
const GROWTH_VECTOR_DIM = 768; // the house embedding dimension

// .finite(): z.number() alone accepts +/-Infinity — the Pydantic side 422s any
// non-finite component (it would silently poison every cosine), so the mirror must too.
const growthVector = z.array(z.number().finite()).length(GROWTH_VECTOR_DIM);

export const GrowthPhraseSchema = z.object({
  id: z.string(),
  phrase: z.string(), // ALREADY pseudonymized at rest (SG-1)
  count: z.number().int().min(1),
  vector: growthVector,
});
export type GrowthPhrase = z.infer<typeof GrowthPhraseSchema>;

export const GrowthAnchorSchema = z.object({
  skill_id: z.string(), // the CLOSED id space — the only id a proposal may carry (SG-3)
  vector: growthVector,
});
export type GrowthAnchor = z.infer<typeof GrowthAnchorSchema>;

export const GrowthClusterInputSchema = z.object({
  domain_id: z.string(),
  phrases: z.array(GrowthPhraseSchema).max(500), // request caps == Pydantic max_length
  anchors: z.array(GrowthAnchorSchema).max(5000),
  min_cluster_size: z.number().int().min(1).nullable().default(null),
  min_total_count: z.number().int().min(1).nullable().default(null),
  cluster_threshold: z.number().min(0).max(1).nullable().default(null),
  band_low: z.number().min(0).max(1).nullable().default(null),
  floor: z.number().min(0).max(1).nullable().default(null),
});
export type GrowthClusterInput = z.infer<typeof GrowthClusterInputSchema>;

// kind=alias → skill_id set (ALWAYS one of the request's anchors — SG-3);
// kind=provisional_skill → skill_id null (NO id is minted here — SG-5).
export const GrowthProposalSchema = z.object({
  kind: z.enum(["alias", "provisional_skill"]),
  skill_id: z.string().nullable().default(null),
  leader_phrase: z.string(),
  member_ids: z.array(z.string()),
  member_phrases: z.array(z.string()),
  total_count: z.number().int(),
  nearest_skill_id: z.string().nullable().default(null),
  nearest_score: z.number().nullable().default(null),
  note: z.string().nullable().default(null),
});
export type GrowthProposal = z.infer<typeof GrowthProposalSchema>;

export const GrowthClusterOutputSchema = z.object({
  proposals: z.array(GrowthProposalSchema),
  phrases_in: z.number().int().nonnegative(),
  clusters_total: z.number().int().nonnegative(),
  clusters_eligible: z.number().int().nonnegative(),
  skipped_below_guards: z.number().int().nonnegative(),
});
export type GrowthClusterOutput = z.infer<typeof GrowthClusterOutputSchema>;

// ---------------------------------------------------------------------------
// Offline skill re-tag plan (ADR-0030 / TAX-9 — pure compute, dry-run first) —
// mirrors contracts.py. The db-side runner (packages/db retag-skills.ts) supplies
// the skill.replaced_by crosswalk + affected rows to /skills/retag-plan and applies
// the returned changes only under --apply. row_ref is an opaque row uuid — no PII.
// ---------------------------------------------------------------------------
export const RetagCrosswalkEntrySchema = z.object({
  deprecated_id: z.string(),
  replaced_by: z.string(),
});
export type RetagCrosswalkEntry = z.infer<typeof RetagCrosswalkEntrySchema>;

export const RetagRowSchema = z.object({
  row_ref: z.string(),
  skill_ids: z.array(z.string()).max(100), // caps == Pydantic max_length
});
export type RetagRow = z.infer<typeof RetagRowSchema>;

export const RetagPlanInputSchema = z.object({
  crosswalk: z.array(RetagCrosswalkEntrySchema).max(1000),
  rows: z.array(RetagRowSchema).max(5000),
});
export type RetagPlanInput = z.infer<typeof RetagPlanInputSchema>;

export const RetagResolvedEntrySchema = z.object({
  deprecated_id: z.string(),
  terminal_id: z.string(),
  hops: z.number().int().min(1),
});
export type RetagResolvedEntry = z.infer<typeof RetagResolvedEntrySchema>;

export const RetagChangeSchema = z.object({
  row_ref: z.string(),
  before: z.array(z.string()),
  after: z.array(z.string()),
});
export type RetagChange = z.infer<typeof RetagChangeSchema>;

export const RetagPlanOutputSchema = z.object({
  resolved: z.array(RetagResolvedEntrySchema),
  dropped: z.array(z.string()), // crosswalk ids on a CYCLE — fail-safe, not re-tagged
  changes: z.array(RetagChangeSchema),
  rows_in: z.number().int().nonnegative(),
  rows_changed: z.number().int().nonnegative(),
});
export type RetagPlanOutput = z.infer<typeof RetagPlanOutputSchema>;
